crates/shared/src/interaction.rs

```diff
@@ -0,0 +1,46 @@
+use ethcontract::Bytes;
+use model::interaction::InteractionData as InteractionDataFromOrder;
+use primitive_types::{H160, U256};
+use serde::{Deserialize, Serialize};
+
+pub trait Interaction: std::fmt::Debug + Send + Sync {
+    // TODO: not sure if this should return a result.
+    // Write::write returns a result but we know we write to a vector in memory so we know it will
+    // never fail. Then the question becomes whether interactions should be allowed to fail encoding
+    // for other reasons.
+    fn encode(&self) -> Vec<EncodedInteraction>;
+}
+
+pub type EncodedInteraction = (
+    H160,           // target
+    U256,           // value
+    Bytes<Vec<u8>>, // callData
+);
+
+impl Interaction for EncodedInteraction {
+    fn encode(&self) -> Vec<EncodedInteraction> {
+        vec![self.clone()]
+    }
+}
+
+#[derive(Eq, PartialEq, Clone, Debug, Hash, Default, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct InteractionData {
```

Comment:
Instead of needing an additional struct, we should be able to implement:
```rust
impl Interaction for model::interaction::InteractionData {
    // ...
}
```

This should be allowed in the `shared` crate because the trait is introduced in that crate.